### DISCUSSION TITLE
Allow colors for individual parts of `sidebar_format`

### DISCUSSION THREAD

**Discussion by whitney-cumber** [neomutt/neomutt#4517] [Development]
## What would you like to be able to do?

Color individual parts of an entry in the sidebar, e.g. the number of flagged messages.

## What needs to change in NeoMutt?

A can imaging two a mechanisms:

* similar to color object `index_FOO` where individual expands can be coloured
* similar to color object `status` which allows an additional regex to match

Dreaming even further on one could work towards unifying that syntax by introducing a new expando sequence in all the `_format` variables which allows to name a group which a color command can match.

Example:
`%gr?<name>{` and `%gr}` are the new expando sequences.

```
set index_format = "%a %s %* %gr?foo{[%y]%gr}"
color index red blue foo
```
Note that not only `%y` is coloured but also the brackets `[%y]`.

This can even be extended to

```
set index_format = "%a %s %* %gr?foo{[%y]%gr}"
color index red blue foo '\d*(K)' 1 ~N
```

Which would only colour the letter `K` after digits inside of the *foo*-group but only for new mail. In this generality it can replace all the current colour schemes (maybe the command should be called `color2` for better backwards compatibility)

## Have you tried any alternative methods?

No.

**Comment by flatcap** [comment]
This is a great idea, though that final scheme is a bit of a mind-blower!
All the machinery is already there, so it shouldn't be too hard to implement.

What's your current `sidebar_format`?
Would a `color status` model be enough?

Mine is:
```
set sidebar_format = "%D%?F? [%F]?%* %?N?%N/?%?S?%S?"
```

so this might work:
```
# Flagged messages
color sidebar white red '\[[0-9]+\]'

# Just the number
color sidebar white red '\[([0-9]+)\]' 1
```

or

```
# New messages
color sidebar brightyellow default '([0-9]+)/[0-9]+$' 1
```

I _think_ :-)

**Comment by whitney-cumber** [comment]
> This is a great idea, though that final scheme is a bit of a mind-blower!

Yes it is. :-S

> What's your current sidebar_format?

Currently, I have "%B%?F? [%F]?%* %?N?%N/?%S" and I wanted to print the "new mail" number in bold.

> Would a color status model be enough?

For my initial idea yes.  I'm sure at one point someone comes along and wants to have the total size number bold but only for mailboxes which contain new mails.

I think it might be worth thinking about how to combine them.  I'm currently looking at the bazillion `index_foo` color objects adding another bazillion for `sidebar_format`, `attach_format`, ... seems, well, unreasonable.  Better invest some time now to figure out a good API / UI for the purpose instead of adding adhoc and little by little some functionality.

Hence, my initial idea that the user can specify the parts of the format string they later want to colour.  This resulted in the suggested `%gr(` expandos.

Later I noticed the following: Sometimes the colour should not only be depend the type (e.g. the sender of an mail) but also on its value (e.g. the sender is my best friend).  Hence, the need to also support regex in a `color` command (like for status).

However, this still does not incorporate the selection of the colour based on a property which is not visibly reflected in the line to format, e.g. if the mail is unread (that CAN be part of it, but if i try to colour code it then the visible `N` marker is no more).  Hence the addition of the pattern.

Maybe something like
```
colour index red blue group=foo regex=[0-9]+ regex_match=1 pattern=~N
```
would be more where everything after the colour is an optional condition and to colour it every condition must be met.
```
colour index red blue group=foo regex=[0-9]+ regex_match=1 pattern=~N
colour index red blue pattern=~N regex=[0-9]+ regex_match=1 group=foo # the same as before 
colour index red blue group=foo  # must be inside the group foo
colour index red blue regex=[0-9]+ # must match the regex
colour index red blue pattern=~N  # must be a new mail
```
This also allows to add more conditions later (if it is every needed).

You looked into such stuff longer than I have.  Maybe there is a simpler method of achieving similar flexibility and / or unifying the format and code?

**Comment by sand4rt** [comment]
Did anyone found a way to change the colors for individual parts of the sidebar?
